File: impl/ts/tests/execute_script.test.ts
// S2 executeScript host-API surface (spec/design/session.md §4.2): the multi-statement
// migration/import convenience — split, run each in order, discard rows, return the O(1)
// ScriptSummary. All-or-nothing when Idle, join-when-Open, in-script transaction control 0A000.
// Host-API behaviors the single-statement corpus cannot call (CLAUDE.md §10). Mirrors
// impl/rust/tests/execute_script.rs.

import assert from "node:assert/strict";
import { test } from "node:test";
import { Database, EngineError, execute, intValue } from "../src/lib.ts";

function code(fn: () => unknown): string {
  try {
    fn();
  } catch (e) {
    if (e instanceof EngineError) return e.code();
    throw e;
  }
  throw new Error("expected an error, got none");
}

function countRows(db: Database): unknown {
  const out = execute(db, "SELECT count(*) FROM t");
  if (out.kind !== "query") throw new Error("expected a query result");
  return out.rows;
}

test("script summary counts and commits atomically when Idle", () => {
  const db = new Database();
  const summary = db.executeScript(
    `CREATE TABLE t (id i32 PRIMARY KEY, v i32);
     INSERT INTO t VALUES (1, 10);
     INSERT INTO t VALUES (2, 20), (3, 30);
     UPDATE t SET v = v + 1 WHERE id >= 2;
     DELETE FROM t WHERE id = 1;`,
  );
  assert.strictEqual(summary.statementsRun, 5);
  assert.strictEqual(summary.rowsAffectedTotal, 1 + 2 + 2 + 1); // DDL contributes 0
  assert.ok(summary.cost > 0n);
  assert.strictEqual(db.status(), "Idle");
  assert.deepStrictEqual(countRows(db), [[intValue(2n)]]); // ids 2 and 3 remain
});

test("script is all-or-nothing on error", () => {
  const db = new Database();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  assert.strictEqual(
    code(() =>
      db.executeScript(
        "INSERT INTO t VALUES (1); INSERT INTO t VALUES (2); INSERT INTO t VALUES (1)",
      ),
    ),
    "23505",
  );
  assert.strictEqual(db.status(), "Idle");
  assert.deepStrictEqual(countRows(db), [[intValue(0n)]]); // nothing committed
});

test("script SELECT rows are discarded but the statement is counted", () => {
  const db = new Database();
  const summary = db.executeScript(
    `CREATE TABLE t (id i32 PRIMARY KEY);
     INSERT INTO t VALUES (1), (2);
     SELECT * FROM t;`,
  );
  assert.strictEqual(summary.statementsRun, 3);
  assert.strictEqual(summary.rowsAffectedTotal, 2); // only the INSERT
  assert.deepStrictEqual(countRows(db), [[intValue(2n)]]);
});

test("empty script is a no-op success", () => {
  const db = new Database();
  const summary = db.executeScript("  -- just a comment\n /* and a block */ ;;; ");
  assert.deepStrictEqual(summary, { statementsRun: 0, rowsAffectedTotal: 0, cost: 0n });
  assert.strictEqual(db.status(), "Idle");
});

test("in-script transaction control is 0A000 and rolls the run back", () => {
  const db = new Database();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  for (const script of [
    "INSERT INTO t VALUES (1); COMMIT; INSERT INTO t VALUES (2)",
    "INSERT INTO t VALUES (1); BEGIN; INSERT INTO t VALUES (2)",
    "INSERT INTO t VALUES (1); ROLLBACK",
  ]) {
    assert.strictEqual(
      code(() => db.executeScript(script)),
      "0A000",
      script,
    );
    assert.strictEqual(db.status(), "Idle");
    assert.deepStrictEqual(countRows(db), [[intValue(0n)]], script); // wrapper rolled back
  }
});

test("script joins an open transaction without committing", () => {
  const db = new Database();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  execute(db, "BEGIN");
  const summary = db.executeScript("INSERT INTO t VALUES (1); INSERT INTO t VALUES (2)");
  assert.strictEqual(summary.statementsRun, 2);
  assert.strictEqual(db.status(), "Open"); // NOT auto-committed — the caller's block stays open
  assert.deepStrictEqual(countRows(db), [[intValue(2n)]]); // visible inside the block
  execute(db, "ROLLBACK");
  assert.deepStrictEqual(countRows(db), [[intValue(0n)]]); // the caller rolled it back
});

test("script error inside an open transaction leaves it Failed for the caller", () => {
  const db = new Database();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  execute(db, "BEGIN");
  assert.strictEqual(
    code(() => db.executeScript("INSERT INTO t VALUES (1); INSERT INTO t VALUES (1)")),
    "23505",
  );
  assert.strictEqual(db.status(), "Failed"); // executeScript does NOT roll back a tx it doesn't own
  execute(db, "ROLLBACK");
  assert.strictEqual(db.status(), "Idle");
});

test("additional session runs a script via the swap", () => {
  const db = new Database();
  execute(db, "CREATE TABLE t (id i32 PRIMARY KEY)");
  const s = db.newSession();
  const summary = s.executeScript(db, "INSERT INTO t VALUES (1); INSERT INTO t VALUES (2)");
  assert.strictEqual(summary.statementsRun, 2);
  assert.deepStrictEqual(countRows(db), [[intValue(2n)]]);
  assert.strictEqual(db.status(), "Idle");
});

File: impl/go/execute_script_test.go
package jed

// S2 ExecuteScript host-API surface (spec/design/session.md §4.2): the multi-statement
// migration/import convenience — split, run each in order, discard rows, return the O(1)
// ScriptSummary. All-or-nothing when Idle, join-when-Open, in-script transaction control 0A000.
// Host-API behaviors the single-statement corpus cannot call (CLAUDE.md §10). Mirrors
// impl/rust/tests/execute_script.rs.

import "testing"

func scriptCount(t *testing.T, db *Engine) int64 {
	t.Helper()
	out, err := Execute(db, "SELECT count(*) FROM t")
	if err != nil {
		t.Fatalf("count: %v", err)
	}
	return out.Rows[0][0].Int
}

func TestScriptSummaryCountsAndCommitsAtomicallyWhenIdle(t *testing.T) {
	db := NewEngine()
	summary, err := db.ExecuteScript(
		`CREATE TABLE t (id i32 PRIMARY KEY, v i32);
		 INSERT INTO t VALUES (1, 10);
		 INSERT INTO t VALUES (2, 20), (3, 30);
		 UPDATE t SET v = v + 1 WHERE id >= 2;
		 DELETE FROM t WHERE id = 1;`,
	)
	if err != nil {
		t.Fatalf("ExecuteScript: %v", err)
	}
	if summary.StatementsRun != 5 {
		t.Fatalf("StatementsRun: want 5, got %d", summary.StatementsRun)
	}
	if summary.RowsAffectedTotal != 1+2+2+1 { // insert+insert+update+delete; DDL = 0
		t.Fatalf("RowsAffectedTotal: want 6, got %d", summary.RowsAffectedTotal)
	}
	if summary.Cost <= 0 {
		t.Fatalf("Cost: want > 0, got %d", summary.Cost)
	}
	if db.Status() != TxIdle {
		t.Fatalf("status: want Idle, got %v", db.Status())
	}
	if n := scriptCount(t, db); n != 2 {
		t.Fatalf("count: want 2, got %d", n)
	}
}

func TestScriptIsAllOrNothingOnError(t *testing.T) {
	db := NewEngine()
	sessExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY)")
	_, err := db.ExecuteScript("INSERT INTO t VALUES (1); INSERT INTO t VALUES (2); INSERT INTO t VALUES (1)")
	if got := sessCode(t, err); got != "23505" {
		t.Fatalf("want 23505, got %s", got)
	}
	if db.Status() != TxIdle {
		t.Fatalf("status: want Idle, got %v", db.Status())
	}
	if n := scriptCount(t, db); n != 0 {
		t.Fatalf("count: want 0 (all-or-nothing), got %d", n)
	}
}

func TestScriptSelectRowsDiscardedButStatementCounted(t *testing.T) {
	db := NewEngine()
	summary, err := db.ExecuteScript(
		`CREATE TABLE t (id i32 PRIMARY KEY);
		 INSERT INTO t VALUES (1), (2);
		 SELECT * FROM t;`,
	)
	if err != nil {
		t.Fatalf("ExecuteScript: %v", err)
	}
	if summary.StatementsRun != 3 {
		t.Fatalf("StatementsRun: want 3, got %d", summary.StatementsRun)
	}
	if summary.RowsAffectedTotal != 2 { // only the INSERT; the SELECT contributes 0
		t.Fatalf("RowsAffectedTotal: want 2, got %d", summary.RowsAffectedTotal)
	}
	if n := scriptCount(t, db); n != 2 {
		t.Fatalf("count: want 2, got %d", n)
	}
}

func TestEmptyScriptIsANoOpSuccess(t *testing.T) {
	db := NewEngine()
	summary, err := db.ExecuteScript("  -- just a comment\n /* and a block */ ;;; ")
	if err != nil {
		t.Fatalf("ExecuteScript: %v", err)
	}
	if summary != (ScriptSummary{}) {
		t.Fatalf("want zero summary, got %+v", summary)
	}
	if db.Status() != TxIdle {
		t.Fatalf("status: want Idle, got %v", db.Status())
	}
}

func TestInScriptTransactionControlIsFeatureNotSupported(t *testing.T) {
	db := NewEngine()
	sessExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY)")
	for _, script := range []string{
		"INSERT INTO t VALUES (1); COMMIT; INSERT INTO t VALUES (2)",
		"INSERT INTO t VALUES (1); BEGIN; INSERT INTO t VALUES (2)",
		"INSERT INTO t VALUES (1); ROLLBACK",
	} {
		_, err := db.ExecuteScript(script)
		if got := sessCode(t, err); got != "0A000" {
			t.Fatalf("%q: want 0A000, got %s", script, got)
		}
		if db.Status() != TxIdle {
			t.Fatalf("%q: status want Idle, got %v", script, db.Status())
		}
		if n := scriptCount(t, db); n != 0 {
			t.Fatalf("%q: count want 0 (wrapper rolled back), got %d", script, n)
		}
	}
}

func TestScriptJoinsAnOpenTransactionWithoutCommitting(t *testing.T) {
	db := NewEngine()
	sessExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY)")
	sessExec(t, db, "BEGIN")
	summary, err := db.ExecuteScript("INSERT INTO t VALUES (1); INSERT INTO t VALUES (2)")
	if err != nil {
		t.Fatalf("ExecuteScript: %v", err)
	}
	if summary.StatementsRun != 2 {
		t.Fatalf("StatementsRun: want 2, got %d", summary.StatementsRun)
	}
	if db.Status() != TxOpen { // NOT auto-committed — the caller's block stays open
		t.Fatalf("status: want Open, got %v", db.Status())
	}
	if n := scriptCount(t, db); n != 2 {
		t.Fatalf("count inside block: want 2, got %d", n)
	}
	sessExec(t, db, "ROLLBACK")
	if n := scriptCount(t, db); n != 0 {
		t.Fatalf("count after rollback: want 0, got %d", n)
	}
}

func TestScriptErrorInsideOpenTransactionLeavesItFailed(t *testing.T) {
	db := NewEngine()
	sessExec(t, db, "CREATE TABLE t (id i32 PRIMARY KEY)")
	sessExec(t, db, "BEGIN")
	_, err := db.ExecuteScript("INSERT INTO t VALUES (1); INSERT INTO t VALUES (1)")
	if got := sessCode(t, err); got != "23505" {
		t.Fatalf("want 23505, got %s", got)
	}
	if db.Status() != TxFailed { // execute_script does NOT roll back a tx it does not own
		t.Fatalf("status: want Failed, got %v", db.Status())
	}
	sessExec(t, db, "ROLLBACK")
	if db.Status() != TxIdle {
		t.Fatalf("status: want Idle, got %v", db.Status())
	}
}

func TestAdditionalSessionRunsAScriptOverTheSharedCore(t *testing.T) {
	// ExecuteScript on an ADDITIONAL session (§2.1/§2.4) shares committed storage through the Database
	// core and commits the run all-or-nothing — another session sees it.
	db := NewDatabase()
	a := db.Session(SessionOptions{})
	if _, err := a.Execute("CREATE TABLE t (id i32 PRIMARY KEY)", nil); err != nil {
		t.Fatal(err)
	}
	s := db.Session(SessionOptions{})
	summary, err := s.ExecuteScript("INSERT INTO t VALUES (1); INSERT INTO t VALUES (2)")
	if err != nil {
		t.Fatalf("ExecuteScript: %v", err)
	}
	if summary.StatementsRun != 2 {
		t.Fatalf("StatementsRun: want 2, got %d", summary.StatementsRun)
	}
	out, err := a.Execute("SELECT count(*) FROM t", nil)
	if err != nil {
		t.Fatal(err)
	}
	if out.Rows[0][0].Int != 2 {
		t.Fatalf("count: want 2, got %v", out.Rows[0][0])
	}
	if a.Status() != TxIdle {
		t.Fatalf("status: want Idle, got %v", a.Status())
	}
}
